src/riot-rs-threads/src/arch/xtensa.rs

```diff
@@ -0,0 +1,141 @@
+        thread.data.A4 = cleanup as u32;
+        thread.data.PC = func as u32;
+
+        // Copied from esp-wifi::preempt::preempt_xtensa
+
+        // For windowed ABI set WOE and CALLINC (pretend task was 'call4'd).
+        thread.data.PS = 0x00040000 | (1 & 3) << 16;
+    }
+
+    fn start_threading() {
+        interrupt::disable(esp_hal::Cpu::ProCpu, Interrupt::FROM_CPU_INTR1);
+        Self::schedule();
+        // Panics if `FROM_CPU_INTR1` is among `esp_hal::interrupt::RESERVED_INTERRUPTS`,
+        // which isn't the case.
+        interrupt::enable(Interrupt::FROM_CPU_INTR1, interrupt::Priority::min()).unwrap();
+    }
+}
+
+const fn default_trap_frame() -> TrapFrame {
+    TrapFrame {
+        PC: 0,
+        PS: 0,
+        A0: 0,
+        A1: 0,
+        A2: 0,
+        A3: 0,
+        A4: 0,
+        A5: 0,
+        A6: 0,
+        A7: 0,
+        A8: 0,
+        A9: 0,
+        A10: 0,
+        A11: 0,
+        A12: 0,
+        A13: 0,
+        A14: 0,
+        A15: 0,
+        SAR: 0,
+        EXCCAUSE: 0,
+        EXCVADDR: 0,
+        LBEG: 0,
+        LEND: 0,
+        LCOUNT: 0,
+        THREADPTR: 0,
+        SCOMPARE1: 0,
+        BR: 0,
+        ACCLO: 0,
+        ACCHI: 0,
+        M0: 0,
+        M1: 0,
+        M2: 0,
+        M3: 0,
+    }
+}
+
+/// Handler for software interrupt 0, which we use for context switching.
+#[allow(non_snake_case)]
+#[no_mangle]
+extern "C" fn FROM_CPU_INTR1(trap_frame: &mut TrapFrame) {
+    unsafe {
+        // clear FROM_CPU_INTR1
+        (&*SYSTEM::PTR)
+            .cpu_intr_from_cpu_1()
+            .modify(|_, w| w.cpu_intr_from_cpu_1().clear_bit());
+
+        sched(trap_frame)
+    }
+}
+
+/// Probes the runqueue for the next thread and switches context if needed.
+///
+/// # Safety
+///
+/// This method might switch the current register state that is saved in the
+/// `trap_frame`.
+/// It should only be called from inside the trap handler that is responsible for
+/// context switching.
+unsafe fn sched(trap_frame: &mut TrapFrame) {
+    loop {
+        if THREADS.with_mut(|mut threads| {
+            let Some(next_pid) = threads.runqueue.get_next() else {
+                return false;
+            };
+
+            if let Some(current_pid) = threads.current_pid() {
+                if next_pid == current_pid {
+                    return true;
+                }
+                threads.threads[usize::from(current_pid)].data = *trap_frame;
+            }
+            threads.current_thread = Some(next_pid);
+            *trap_frame = threads.threads[usize::from(next_pid)].data;
+            true
+        }) {
+            break;
+        }
+        // The esp-hal implementation of critical-section doesn't disable all interrupts.
+        // Thus we should release our hold on `THREADS` before we `waiti`, to prevent
+        // that another interrupt handler will try to borrow it while we still have it borrowed.
+        unsafe { core::arch::asm!("waiti 0") };
```

Comment:
We could also disabled all interrupts when accessing `THREADS` on xtensa. `esp-hal` originally disabled all interrupts in their critical section, but changed it to only disable interrupts with prio <=5 so that the debugging at interrupt level 6 still works (see https://github.com/esp-rs/esp-hal/pull/360). 
So if we'd change it, debugging wouldn't work inside critical sections anymore I think.
On the other hand, it can happen right now that an `esp_hal::embassy` interrupt interrupts a running thread that is currently accessing `THREADS`. If that interrupt is then also accessing `THREADS` (e.g. because it's a `_pender`), it panics.
But the closures for accessing `THREADS` are very short, so I don't know.

Either way the `waiti` here still would need to be outside of the closure, because IIUC it only returns if an interrupt **at the currently enabled level** occurs. If all interrupts are disabled, it will never return.

Reply:
if that `prio <= 5` refers to the `Priority` type, this might be fine, see: https://github.com/kaspar030/esp-hal/blob/23d22e32784e8c9b418a4fb4e316397192e6e32b/esp-hal/src/interrupt/xtensa.rs#L355-L356. Also, `Priority::max() == Priority3`. So embassy (or any other Rust) code will probably not be able to access `THREADS`

Reply:
then again, we might prefer stalling at debug time over possible panic at non-debug time ...

Reply:
Okay so seems like interrupt level up until level 7 is still possible, or at least configured for xtensa-lx-rt in https://github.com/esp-rs/esp-hal/blob/3bb49b049a2d95dd2f6fc1b6710941f5938c1db7/xtensa-lx-rt/config/xtensa_esp32s3.toml#L190-L191.
But looking through that list only the debug interrupt has an interrupt level > 5, and the NMI interrupt / INT14, but that one can't be masked anyway :sweat_smile:  And `CpuInterrupt::Interrupt14NmiPriority7` is among the interrupts that are mapped to `None` in your above code.
So I don't think it'll be an issue for us here.